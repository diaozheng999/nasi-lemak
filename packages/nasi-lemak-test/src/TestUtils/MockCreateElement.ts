/**
 * MockCreateElement.ts
 * @author Diao Zheng
 * @file Mocking React.createElement for testing
 */

import { Disposable, Types, Unique } from "nasi-lemak";
import { Attributes, FunctionComponent, ReactElement, ReactNode } from "react";
import { ConcurrentSideEffectChain } from "../EffectChains";
import {
  __internal_getCurrentExecutor,
  __internal_resetCurrentHookCount,
  __internal_setCurrentExecutor,
} from "./CurrentExecutor";

export function MockCreateElement(ReactActual: any) {
  return <P>(
    type: FunctionComponent<P>,
    props?: Attributes & P | null,
    ...children: ReactNode[]
  ): ReactElement<P> => {
    // in JavaScript a class constructor is typeof function. So, we need to
    // keep in mind both class components as well as function components
    if (typeof type === "function") {
      // checks if it's a class component.
      if (type.prototype.isReactComponent) {
        return ReactActual.createElement(type, props, ...children);
      }

      // it's not a class component. Here we assume that it's a
      // function component
      return ReactActual.createElement((
        ...args: Types.ArgumentTupleType<FunctionComponent<P>>
      ) => {

        // we hijack useState to create a memoised executor chain. This executor
        // will exist for every element.
        const [ executor ] = ReactActual.useState(
          () => {
            const parentExecutor = __internal_getCurrentExecutor();
            const currentExecutor = new ConcurrentSideEffectChain(
              parentExecutor,
              new Unique(
                (type.displayName ?? type.name ?? "FunctionComponent") +
                "_Hook",
              ),
              true,
            );
            parentExecutor.enqueue(currentExecutor);
            return currentExecutor;
          },
        );

        const previousExecutor = __internal_setCurrentExecutor(executor);
        __internal_resetCurrentHookCount();

        // execute the real FunctionComponent constructor
        const returnValue = type(...args);

        // we will immediate remove the child hook if none of the
        // SideEffectChain hooks have been committed.
        ReactActual.useLayoutEffect(() => {
          const count = __internal_resetCurrentHookCount();

          if (count) {
            return () => Disposable.dispose(executor);
          } else {
            Disposable.dispose(executor);
            return;
          }
        }, []);

        __internal_setCurrentExecutor(previousExecutor);

        return returnValue;

      }, props, ...children);
    } else {
      // here, it's a built-in or exotic component (i.e. symbols with $$typeof)
      return ReactActual.createElement(type, props, ...children);
    }
  };

}
