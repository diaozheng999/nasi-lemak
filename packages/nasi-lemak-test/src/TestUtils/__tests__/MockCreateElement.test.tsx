/**
 * MockCreateElement.test.ts
 * @author Diao Zheng
 * @file Tests to make sure React components are created properly
 */

// tslint:disable-next-line:no-reference
/// <reference path="../../../node_modules/@types/jest/index.d.ts" />

import renderer from "react-test-renderer";
import { MockCreateElement } from "../MockCreateElement";

jest.doMock("react", () => {
  const ReactActual = jest.requireActual("react");
  return {
    ...ReactActual,
    createElement: MockCreateElement(ReactActual),
  };
});

import React, { Suspense } from "react";

class TestClassComponent extends React.Component<any> {
  public render() {
    return React.createElement("__builtin__", this.props);
  }
}

function TestFunctionComponent() {
  return React.createElement("__builtin__", {});
}

function fetcher() {
  return {
    read() {
      return 5;
    },
  };
}

const resource = fetcher();

function TestAsyncComponent() {
  const p = resource.read();
  return <TestClassComponent value={p} />;
}

const TestContext = React.createContext("TestContext");

const TestExoticComponent = React.forwardRef(
  (props, ref: React.Ref<TestClassComponent>) => {
    return <TestClassComponent {...props} ref={ref} />;
  },
);

test("function component", () => {
  const tree = renderer.create(<TestFunctionComponent />);
  expect(tree.toJSON()).toMatchSnapshot();
});

test("class component", () => {
  const tree = renderer.create(<TestClassComponent />);
  expect(tree.toJSON()).toMatchSnapshot();
});

test("builtin", () => {
  const tree = renderer.create(<h6 />);
  expect(tree.toJSON()).toMatchSnapshot();
});

test("context producer", () => {
  const tree = renderer.create(
    <TestContext.Provider value="boo">
      <TestFunctionComponent />
    </TestContext.Provider>,
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

test("context consumer", () => {
  const tree = renderer.create(
    <TestContext.Consumer>
      {(val) => <TestClassComponent value={val} />}
    </TestContext.Consumer>,
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

test("fragment", () => {
  const tree = renderer.create(
    <>
      <TestFunctionComponent />
      <TestClassComponent />
    </>,
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

test("exotic", () => {
  const tree = renderer.create(<TestExoticComponent />);
  expect(tree.toJSON()).toMatchSnapshot();
});

test("suspense", () => {
  const tree = renderer.create(
    <Suspense fallback={<h1/>}>
      <TestAsyncComponent />
    </Suspense>,
  );
  expect(tree.toJSON()).toMatchSnapshot();
});
