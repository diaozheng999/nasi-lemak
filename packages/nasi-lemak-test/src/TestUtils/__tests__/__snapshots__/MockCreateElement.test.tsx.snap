// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`builtin 1`] = `<h6 />`;

exports[`class component 1`] = `<__builtin__ />`;

exports[`context consumer 1`] = `
<__builtin__
  value="TestContext"
/>
`;

exports[`context producer 1`] = `<__builtin__ />`;

exports[`exotic 1`] = `<__builtin__ />`;

exports[`fragment 1`] = `
Array [
  <__builtin__ />,
  <__builtin__ />,
]
`;

exports[`function component 1`] = `<__builtin__ />`;

exports[`suspense 1`] = `
<__builtin__
  value={5}
/>
`;
